### PR TITLE
Account for possible cache invalidation on a different thread in assert.

### PR DESCRIPTION
**Pull request by 333fred**
Attempt at fixing https://github.com/dotnet/roslyn/issues/59395. Reviewing the codepath I don't see any other obvious points that could cause the assert to fail, so if it continues to reproduce after this change we can look at further logging to determine the root cause.

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 1]
```diff
@@ -1169,7 +1169,9 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
             {
                 val = CommonGetTypeByMetadataName(fullyQualifiedMetadataName);
                 var result = _getTypeCache.TryAdd(fullyQualifiedMetadataName, val);
-                Debug.Assert(result || (_getTypeCache.TryGetValue(fullyQualifiedMetadataName, out var addedType) && ReferenceEquals(addedType, val)));
+                Debug.Assert(result
+                 || !_getTypeCache.TryGetValue(fullyQualifiedMetadataName, out var addedType) // Could fail if the type was already evicted from the cache
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:1173]: >// Could fail if the type was already evicted from the cache [](http://example.com/codeflow?start=94&length=60)

Technically, I think it is possible that the type got evicted and then a different instance got added by other threads.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:1173]: Or if we add a type we should always get the same instance?

Comment (333fred) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:1173]: > Or if we add a type we should always get the same instance?

I believe it should always be the same instance. This API effectively just walks through every assembly and looks for the member symbols from the assembly, and (in C#) gets the public version. Those should not be changing.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 1)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

**Pull request closed without merging**